Add support for JSON formatted logs

mriedem opened the pull request:
This adds a new extra package dependency on `json-logging`
and an environment variable, which when set to "true" regardless
of case, will try to use the json-logging non-web app log formatter.
A `NotebookApp.log_json` config option is also added which gets
its default value from the `JUPYTER_ENABLE_JSON_LOGGING`
environment variable.

If the `json-logging` package is not installed but `log_json` is True,
something like this will be logged but it will not crash the application:

```
$ JUPYTER_ENABLE_JSON_LOGGING=true jupyter notebook
Unable to use json logging due to missing packages. Run "pip install notebook[json-logging]" to fix.
Traceback (most recent call last):
  File "/home/osboxes/jupyter/notebook/notebook/notebookapp.py", line 144, in <module>
    import json_logging
ModuleNotFoundError: No module named 'json_logging'
```

With this working you get logs like this:

```
{"written_at": "2020-10-07T21:10:51.606Z", "written_ts": 1602105051606265000, "msg": "404 GET /nbextensions/widgets/notebook/js/extension.js (127.0.0.1) 9.26ms referer=http://localhost:8888/notebooks/Untitled.ipynb", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "WARNING", "module": "log", "line_no": 49}
{"written_at": "2020-10-07T21:10:54.443Z", "written_ts": 1602105054443309000, "msg": "Starting buffering for f260ddd7-938c-42d0-ac3b-455bea76694f:49f30b53fc4b4ec6a8f2fb748a171613", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "kernelmanager", "line_no": 222}
{"written_at": "2020-10-07T21:10:54.446Z", "written_ts": 1602105054446264000, "msg": "Kernel shutdown: f260ddd7-938c-42d0-ac3b-455bea76694f", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "multikernelmanager", "line_no": 201}
```

Web app requests get logged with separate properties like this:

```json
{"written_at": "2020-10-09T20:59:52.044Z", "written_ts": 1602277192044732000, "msg": "", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "WARNING", "module": "log", "line_no": 50, "status": 404, "method": "GET", "ip": "127.0.0.1", "uri": "/api/kernels/a6dd447b-282a-4e1c-ab65-eb340ad12965/channels?session_id=472907c27ed94e2cbfbae86501f7d159", "request_time": 8.63, "referer": "None"}
```

The nice thing about that is you can filter request logs on certain error
status codes or request times that take longer than some threshold to
setup alerts in your logging infrastructure (ELK, LogDNA, etc).

Closes #5798

mriedem commented:
@sdague @rmoe Here is an idea for a start on JSON log formatting though it might not be super useful if we can't split apart the request-specific parts of the `msg` field.

mriedem commented:
This is the error I get when trying to use `JSONLogWebFormatter`:

```
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib/python3.6/logging/__init__.py", line 994, in emit
    msg = self.format(record)
  File "/usr/lib/python3.6/logging/__init__.py", line 840, in format
    return fmt.format(record)
  File "/home/osboxes/jupyter/notebook/.tox/lib/python3.6/site-packages/json_logging/__init__.py", line 243, in format
    log_object = self._format_log_object(record, request_util=_request_util)
  File "/home/osboxes/jupyter/notebook/.tox/lib/python3.6/site-packages/json_logging/__init__.py", line 341, in _format_log_object
    "correlation_id": request_util.get_correlation_id(within_formatter=True),
AttributeError: 'NoneType' object has no attribute 'get_correlation_id'
```

And it looks like that's because I'm not going through one of the `init_*` framework methods:

https://github.com/bobbui/json-logging-python/blob/1.2.8/json_logging/__init__.py#L132

mriedem commented:
Getting a little further by changing the `log_request` method so I see something like this now:

```json
{"written_at": "2020-10-07T21:52:16.975Z", "written_ts": 1602107536975497000, "msg": "{'status': 404, 'method': 'GET', 'ip': '127.0.0.1', 'uri': '/nbextensions/widgets/notebook/js/extension.js', 'request_time': 1.9807815551757812, 'referer': 'http://localhost:8888/notebooks/test.ipynb'}", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "WARNING", "module": "log", "line_no": 54}
```

But the `msg` is a serialized JSON string now.

mriedem commented on notebook/log.py:
```diff
@@ -46,5 +50,11 @@ def log_request(handler):
         log_method(json.dumps(dict(request.headers), indent=2))
-    log_method(msg.format(**ns))
+    if _enable_json_logs:
+        # FIXME: this still logs the msg as a serialized json string,
```
I wonder if we could use a partial to pass `log` in here:

https://github.com/jupyter/notebook/blob/b617d910521d7ffb5d9099b2307c6026dca4f3da/notebook/notebookapp.py#L269

Because that'd be the `NotebookApp` logger which is using our json log formatter.

mriedem commented on notebook/log.py:
```diff
@@ -46,5 +50,11 @@ def log_request(handler):
         # log all headers if it caused an error
         log_method(json.dumps(dict(request.headers), indent=2))
-    log_method(msg.format(**ns))
+    if _enable_json_logs:
```
Just because this is true doesn't mean that json-logging is installed. If it's not, the request log line ends up looking like this:

```
[W 17:04:57.097 NotebookApp] {'status': 404, 'method': 'GET', 'ip': '127.0.0.1', 'uri': '/nbextensions/widgets/notebook/js/extension.js', 'request_time': 6.743431091308594, 'referer': 'http://localhost:8888/notebooks/Untitled.ipynb?kernel_name=python3'}
```

Also note that with the json logging we're not formatting the request time to 2 decimal places:

`'request_time': 6.743431091308594`

kevin-bates commented:
Thanks Matt.  This is an interesting conundrum on a couple counts.
1. The logging formatter is something used very early - introducing potential chicken/egg situations.
2. The tornado logging is a different facility, yet ideally, we want to leverage the same configurable setting to determine if JSON formatting should be used.

I think these can be resolved in a more application-friendly manner by first introducing a boolean configurable `json_logging` that can be configured via a configuration file using `c.NotebookApp.json_logging = True`.  We'd also want to tie this to a CLI flag (e.g., `--json-logging` such that, when present, it enables JSON formatting).  You should be able to follow the example of `allow_root` for these two details.

Traitlets also provide the ability to _validate_ a config option.  This is where we could see if the `json_logging` package can be imported and, if so, "tickle" the current log_format value after setting the log_formatter_cls.  The reason for this "tickle" is to trigger the [observation that takes place in the traitlets `Application` class](https://github.com/ipython/traitlets/blob/master/traitlets/config/application.py#L209) - which will perform a reset using the formatter_cls value - as you had mentioned.  Modeling the validation after the 'ip' traitlet can get you started here.

We could let the default value for the `json-logging` flag still derive from the env - which is helpful in containerized envs and this could be done via a _default_json_logging() entry that returns configurable's default.  Again, 'ip' uses a defaulting method that you can model this after.

I apologize for these details, but because there were so many unknowns I had, it was quicker just to try things out.  I have some working code - but I'd rather help you navigate this if you don't mind.  Otherwise, just let me know.

mriedem commented:
@kevin-bates thanks for those detailed pointers.

>I'd rather help you navigate this if you don't mind

Yup, no problem and I definitely appreciate the patience with helping me learn here (I don't really have experience with tornado or traitlets, mostly copy/tweak/test when I need to make changes).

mriedem commented:
@kevin-bates I worked on your suggestions a bit this morning and captured the diff [here](https://gist.github.com/mriedem/7a8226ebd3273cfef611b214e136261a) since it's not working. A few notes:

- I added the config option but with name `log_format_json` so that it would show up alongside `log_format` and `log_datefmt` etc in the help output:

```
--NotebookApp.log_datefmt=<Unicode>
    Default: '%Y-%m-%d %H:%M:%S'
    The date format used by logging formatters for %(asctime)s
--NotebookApp.log_format=<Unicode>
    Default: '[%(name)s]%(highlevel)s %(message)s'
    The Logging format template
--NotebookApp.log_format_json=<Bool>
    Default: False
    Set to True to enable JSON formatted logs. Run "pip install notebook[json-
    logging]" to install the required dependent packages. Can also be set using
    the environment variable ENABLE_JSON_LOGGING=true.
--NotebookApp.log_level=<Enum>
    Default: 30
    Choices: (0, 10, 20, 30, 40, 50, 'DEBUG', 'INFO', 'WARN', 'ERROR', 'CRITICAL')
    Set the log level by value or name.
```

I think that's useful for someone looking at logging options to have those grouped together for context.

- I'm assuming I didn't implement `_validate_log_format_json` correctly because it doesn't seem to be getting hit. I've tried running with both `ENABLE_JSON_LOGGING=true jupyter notebook` and `jupyter notebook --NotebookApp.log_format_json=True`. I never see the logging statements I put in `_validate_log_format_json` though, nor do I see the ImportError traceback if I uninstall `json-logging` and try to start the server with `log_format_json=True`. Is there something I'm missing there about when that validation function is supposed to be called?

- I'm also not sure that I'm "tickling" the `log_format` option correctly by doing `self.log_format = self.log_format`.

Thanks again for the help.

kevin-bates commented:
Thanks Matt.  The "tickle" should actually change the format.  In this case, I removed the color-coding attributes.  Here's the code I had put together the other day.  Since you've changed the traitlet's name, that change would need to be reflected, as well as perhaps logging that the import failed during the validation.

Lastly, the convention for envs is that they be prefixed with `JUPYTER_`.

I found no changes necessary other than in notebookapp.py, but would prefer that you confirm the same.
```python
    json_logging = Bool(config=True)

    @default('json_logging')
    def _default_json_logging(self):
        return os.getenv('JUPYTER_ENABLE_JSON_LOGGING', 'false').lower() == 'true'

    @validate('json_logging')
    def _validate_json_logging(self, proposal):
        value = proposal['value']
        if value is True:
            try:
                import json_logging
                self._log_formatter_cls = json_logging.JSONLogFormatter
                self.log_format = u"[%(levelname)1.1s %(asctime)s.%(msecs).03d %(name)s] %(message)s"
            except ImportError:
                value = False
        return value
```

My concern with the change to `log_format_json` is that it implies the value should be some kind of format string - similar to `log_format`, whereas `json_logging` seems more about the _kind_ of logging that will take place.  The concern being that the "format" approach could be confusing.  Yes, I would agree whatever the name is should be grouped near the others as its all the same "topic".

mriedem commented:
> The "tickle" should actually change the format. In this case, I removed the color-coding attributes.

Got it. I'm not totally sure about that solution since if I set `log_format` specifically and this changed it, I'd probably wonder what's going on if the formatting doesn't show up as expected. If there is no better solution we could at least document something about that in the config help.

>the convention for envs is that they be prefixed with JUPYTER_.

Done.

>I found no changes necessary other than in notebookapp.py

I see a bug in my code, I wasn't using `self._log_formatter_cls`.

>My concern with the change to log_format_json is that it implies the value should be some kind of format string

I renamed the config option to simply `log_json` to take the "format" out of it.

----

I'll tinker with this locally and see what I come up with.

I'm also going to see what happens if we call `json_logging.init_non_web(enable_json=True)` from `init_logging` since `init_non_web` does change the formatter for _all_ of the loggers:

https://github.com/bobbui/json-logging-python/blob/1.2.8/json_logging/__init__.py#L153

mriedem commented:
>I'm also going to see what happens if we call json_logging.init_non_web(enable_json=True) from init_logging since init_non_web does change the formatter for all of the loggers

This works and avoids the need to change `log_format` to trigger the `_log_format_changed` observer. Updated the PR.

Still need to figure out how to get `log.log_request` working properly because if you run without the environment variable set it doesn't log the request parameters in json format:

```
{"written_at": "2020-10-09T19:32:51.825Z", "written_ts": 1602271971825341000, "msg": "404 GET /nbextensions/widgets/notebook/js/extension.js (127.0.0.1) 8.94ms referer=http://localhost:8888/notebooks/Untitled.ipynb", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "WARNING", "module": "log", "line_no": 59}
```

and if you do it logs the request parameters as a serialized json string:

```
{"written_at": "2020-10-09T19:34:05.565Z", "written_ts": 1602272045565346000, "msg": "{'status': 404, 'method': 'GET', 'ip': '127.0.0.1', 'uri': '/api/kernels/48ecd2e7-ae9d-4484-8910-70fb46d72491/channels?session_id=5aa3a7dcedf1427e8ff41ae0cb3dae94', 'request_time': 1.8270015716552734, 'referer': 'None'}", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "WARNING", "module": "log", "line_no": 57}
```

mriedem commented on notebook/notebookapp.py:
```diff
@@ -113,6 +114,7 @@
 from ._tz import utcnow, utcfromtimestamp
 from .utils import (
     check_pid,
+    enable_json_logs,
```
This can go away now, we only need to read the env var from the default function.

mriedem commented on notebook/log.py:
```diff
@@ -46,5 +46,8 @@ def log_request(handler):
     if status >= 500 and status != 502:
         # log all headers if it caused an error
         log_method(json.dumps(dict(request.headers), indent=2))
+    # if _enable_json_logs:
```
Oops, this can go away.

mriedem commented:
I'd like to add tests for this but I'm not sure if there is a great way to test it. Maybe something like start the notebook app server with the env var set and verify the logged message can be serialized and then deserialized JSON?

mriedem commented on notebook/log.py:
```diff
@@ -22,29 +22,32 @@ def log_request(handler):
         msg = msg + ' referer={referer}'
     if status >= 500 and status != 502:
         # log all headers if it caused an error
         log_method(json.dumps(dict(request.headers), indent=2))
```
```suggestion
        if log_json:
            log_method("", extra=dict(props=dict(request.headers)))
        else:
            log_method(json.dumps(dict(request.headers), indent=2))
```

Should probably do something like this, I'd need to be able to recreate it though to see what this looks like.

mriedem replied:
Done.

kevin-bates commented on notebook/notebookapp.py:
```diff
@@ -704,6 +707,44 @@ class NotebookApp(JupyterApp):
+                # If configured for json logs and we can't do it, log a hint.
+                # Only log the error once though.
+                if not self._json_logging_import_error_logged:
+                    logging.getLogger(__name__).exception(
```
I think we should just log a warning here rather than including the exception.  This causes unnecessary noise that the message addresses anyway:
```
Unable to use json logging due to missing packages. Run "pip install notebook[json-logging]" to fix.
Traceback (most recent call last):
  File "/opt/anaconda3/envs/notebook-dev/lib/python3.7/site-packages/notebook/notebookapp.py", line 731, in _validate_log_json
    import json_logging
ModuleNotFoundError: No module named 'json_logging'
```

mriedem replied:
Done.

kevin-bates commented on notebook/notebookapp.py:
```diff
@@ -704,6 +707,44 @@ class NotebookApp(JupyterApp):
+            try:
+                import json_logging
+                self._log_formatter_cls = json_logging.JSONLogFormatter
+                # Note that we don't need to trigger _log_format_changed here
```
Let's remove this comment since its relative to our conversation and not really helpful to folks maintaining this code.

mriedem replied:
Done.

mriedem commented on notebook/notebookapp.py:
```diff
@@ -704,6 +708,44 @@ class NotebookApp(JupyterApp):
 
+    _json_logging_import_error_logged = False
+
+    log_json = Bool(False, config=True,
```
Should I write up a small guide in the docs about using this, somewhere like here?

https://github.com/jupyter/notebook/blob/master/docs/source/config_overview.rst#notebook-server

Or is the config option help sufficient?

kevin-bates replied:
I think the help text you've provided is sufficient.  The other options are explicitly listed that I can see, and I don't think everyday users will be wanting JSON logging.  This is more of a large-scale installation thing.

kevin-bates commented on notebook/notebookapp.py:
```diff
@@ -1590,6 +1631,13 @@ def init_configurables(self):
+            # package is missing then _validate_log_json should have set
+            # log_json=False.
+            import json_logging
+            json_logging.init_non_web(enable_json=True)
```
Let's move this into the validation block.  This will also produce json-formatted logging for the initial set of logged statements, like these...
```
{"written_at": "2020-10-09T21:10:49.519Z", "written_ts": 1602277849519721000, "msg": "Searching ['/opt/anaconda3/envs/notebook-dev/bin', '/Users/kbates/.jupyter', '/opt/anaconda3/envs/nb5386/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 179}
{"written_at": "2020-10-09T21:10:49.520Z", "written_ts": 1602277849520456000, "msg": "Looking for jupyter_config in /etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-09T21:10:49.520Z", "written_ts": 1602277849520942000, "msg": "Looking for jupyter_config in /usr/local/etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-09T21:10:49.521Z", "written_ts": 1602277849521220000, "msg": "Looking for jupyter_config in /opt/anaconda3/envs/nb5386/etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-09T21:10:49.521Z", "written_ts": 1602277849521473000, "msg": "Looking for jupyter_config in /Users/kbates/.jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-09T21:10:49.521Z", "written_ts": 1602277849521719000, "msg": "Looking for jupyter_config in /opt/anaconda3/envs/notebook-dev/bin", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-09T21:10:49.523Z", "written_ts": 1602277849523903000, "msg": "Looking for jupyter_notebook_config in /etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-09T21:10:49.524Z", "written_ts": 1602277849524180000, "msg": "Looking for jupyter_notebook_config in /usr/local/etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-09T21:10:49.524Z", "written_ts": 1602277849524445000, "msg": "Looking for jupyter_notebook_config in /opt/anaconda3/envs/nb5386/etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-09T21:10:49.524Z", "written_ts": 1602277849524692000, "msg": "Looking for jupyter_notebook_config in /Users/kbates/.jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-09T21:10:49.526Z", "written_ts": 1602277849526005000, "msg": "Loaded config file: /Users/kbates/.jupyter/jupyter_notebook_config.py", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 578}
{"written_at": "2020-10-09T21:10:49.526Z", "written_ts": 1602277849526326000, "msg": "Looking for jupyter_notebook_config in /opt/anaconda3/envs/notebook-dev/bin", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-09T21:10:49.529Z", "written_ts": 1602277849529904000, "msg": "initializing json logging", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "notebookapp", "line_no": 1636}
{"written_at": "2020-10-09T21:10:49.541Z", "written_ts": 1602277849541371000, "msg": "Paths used for configuration of jupyter_notebook_config: __/etc/jupyter/jupyter_notebook_config.json", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "config_manager", "line_no": 97}
{"written_at": "2020-10-09T21:10:49.542Z", "written_ts": 1602277849542273000, "msg": "Paths used for configuration of jupyter_notebook_config: __/usr/local/etc/jupyter/jupyter_notebook_config.json", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "config_manager", "line_no": 97}
{"written_at": "2020-10-09T21:10:49.543Z", "written_ts": 1602277849543104000, "msg": "Paths used for configuration of jupyter_notebook_config: __/opt/anaconda3/envs/nb5386/etc/jupyter/jupyter_notebook_config.json", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "config_manager", "line_no": 97}
{"written_at": "2020-10-09T21:10:49.543Z", "written_ts": 1602277849543888000, "msg": "Paths used for configuration of jupyter_notebook_config: __/Users/kbates/.jupyter/jupyter_notebook_config.json", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "config_manager", "line_no": 97}
{"written_at": "2020-10-09T21:10:49.587Z", "written_ts": 1602277849587112000, "msg": "It looks like you're running the notebook from source._    If you're working on the Javascript of the notebook, try running__    npm run build:watch__    in another terminal window to have the system incrementally_    watch and build the notebook's JavaScript for you, as you make changes.", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 232}
{"written_at": "2020-10-09T21:10:49.647Z", "written_ts": 1602277849647173000, "msg": "Serving notebooks from local directory: /Users/kbates/notebooks/alice", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2258}
{"written_at": "2020-10-09T21:10:49.647Z", "written_ts": 1602277849647428000, "msg": "Jupyter Notebook 7.0.0.dev0 is running at:", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2258}
{"written_at": "2020-10-09T21:10:49.647Z", "written_ts": 1602277849647604000, "msg": "http://Kevins-MacBook-Pro.local:8888/?token=e3275dfe8643b393967cdd1d181fde6ef60e2f1a100e98d6", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2258}
{"written_at": "2020-10-09T21:10:49.647Z", "written_ts": 1602277849647719000, "msg": " or http://127.0.0.1:8888/?token=e3275dfe8643b393967cdd1d181fde6ef60e2f1a100e98d6", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2258}
{"written_at": "2020-10-09T21:10:49.647Z", "written_ts": 1602277849647819000, "msg": "Kernels will be managed by the Gateway server running at:", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2258}
{"written_at": "2020-10-09T21:10:49.647Z", "written_ts": 1602277849647914000, "msg": "http://yarn-eg-node-1.fyre.ibm.com:8888", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2258}
{"written_at": "2020-10-09T21:10:49.648Z", "written_ts": 1602277849648021000, "msg": "Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2259}
{"written_at": "2020-10-09T21:10:49.648Z", "written_ts": 1602277849648127000, "msg": "Welcome to Project Jupyter! Explore the various tools available and their corresponding documentation. If you are interested in contributing to the platform, please visit the communityresources section at https://jupyter.org/community.html.", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2261}
```

mriedem replied:
Done.

kevin-bates commented on notebook/notebookapp.py:
```diff
@@ -250,9 +251,11 @@ def init_settings(self, jupyter_app, kernel_manager, contents_manager,
             root_dir = '~' + root_dir[len(home):]
 
+        # Use the NotebookApp logger and its formatting for tornado request logging.
+        log_function = functools.partial(log_request, log=log)
```
I'm not finding this to be necessary.  If I restore to the original, I get the same output.

```
This PR:
{"written_at": "2020-10-09T21:11:30.439Z", "written_ts": 1602277890439353000, "msg": "200 GET /api/contents?type=directory&_=1602277851454 (127.0.0.1) 22.49ms", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "log", "line_no": 49}

with changes reverted:
{"written_at": "2020-10-09T21:19:03.578Z", "written_ts": 1602278343578833000, "msg": "200 GET /api/contents?type=directory&_=1602278343433 (127.0.0.1) 12.31ms", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "log", "line_no": 49}
```

kevin-bates commented on notebook/notebookapp.py:
```diff
@@ -250,9 +251,12 @@ def init_settings(self, jupyter_app, kernel_manager, contents_manager,
 
+        # Use the NotebookApp logger and its formatting for tornado request logging.
+        log_function = functools.partial(
+            log_request, log=log, log_json=jupyter_app.log_json)
```
Are you sure this is necessary?  Prior to your very latest change, I had found the original 'partial' didn't change anything.  I'll need to verify this version of the change, but wanted to get my review comments submitted.

mriedem replied:
Without this the `log_request` function would only get the `handler` argument because that's all tornado will pass to it. Nothing blows up but you don't get JSON formatting on what gets logged in `log_request`.

To compare, if I remove the partial and just pass `log_function=log_request` as before we get something like this:

```json
{"written_at": "2020-10-12T16:29:47.157Z", "written_ts": 1602520187157654000, "msg": "404 GET /nbextensions/widgets/notebook/js/extension.js (127.0.0.1) 2.770000ms referer=http://localhost:8888/notebooks/Untitled.ipynb", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "WARNING", "module": "log", "line_no": 52}
```

But with the change and with JSON logging enabled, we get something like this:

```json
{"written_at": "2020-10-12T16:31:16.698Z", "written_ts": 1602520276698143000, "msg": "", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "WARNING", "module": "log", "line_no": 50, "status": 404, "method": "GET", "ip": "127.0.0.1", "uri": "/nbextensions/widgets/notebook/js/extension.js", "request_time": 7.66, "referer": "http://localhost:8888/notebooks/Untitled.ipynb"}
```

In the former, note that the request fields are in the `msg` field, but in the latter the `ns` request fields are split out as specific fields in the log record which is the goal (for me anyway) so you can build queries in a distributed logging system like ELK or LogDNA and filter on things like `request_time:>=30`.

kevin-bates replied:
I see - thank you.

kevin-bates reviewed (commented):
Hi Matt.  This is coming along, but I think still too spread out.  I've added some comments that should help isolate some of these changes.  Thanks.

mriedem reviewed (commented):
Changes coming soon.

mriedem commented on notebook/notebookapp.py:
```diff
@@ -730,15 +730,14 @@ def _validate_log_json(self, proposal):
         if value:
             try:
                 import json_logging
+                self.log.info('initializing json logging')
```
```suggestion
                logging.getLogger(__name__).debug('initializing json logging')
```

kevin-bates commented on notebook/notebookapp.py:
```diff
@@ -704,6 +708,43 @@ class NotebookApp(JupyterApp):
+        if value:
+            try:
+                import json_logging
+                self.log.debug('initializing json logging')
```
I'm finding this is never produced and am thinking it should be removed.

mriedem replied:
How are you running things? Because I see it like this:

```console
(.tox) osboxes@osboxes:~/jupyter/notebook$ JUPYTER_ENABLE_JSON_LOGGING=true jupyter notebook --debug
[D 15:25:58.460 NotebookApp] Searching ['/home/osboxes/jupyter/notebook', '/home/osboxes/.jupyter', '/home/osboxes/jupyter/notebook/.tox/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files
[D 15:25:58.461 NotebookApp] Looking for jupyter_config in /etc/jupyter
[D 15:25:58.461 NotebookApp] Looking for jupyter_config in /usr/local/etc/jupyter
[D 15:25:58.461 NotebookApp] Looking for jupyter_config in /home/osboxes/jupyter/notebook/.tox/etc/jupyter
[D 15:25:58.461 NotebookApp] Looking for jupyter_config in /home/osboxes/.jupyter
[D 15:25:58.461 NotebookApp] Looking for jupyter_config in /home/osboxes/jupyter/notebook
[D 15:25:58.461 NotebookApp] Looking for jupyter_notebook_config in /etc/jupyter
[D 15:25:58.461 NotebookApp] Looking for jupyter_notebook_config in /usr/local/etc/jupyter
[D 15:25:58.462 NotebookApp] Looking for jupyter_notebook_config in /home/osboxes/jupyter/notebook/.tox/etc/jupyter
[D 15:25:58.462 NotebookApp] Looking for jupyter_notebook_config in /home/osboxes/.jupyter
[D 15:25:58.462 NotebookApp] Looking for jupyter_notebook_config in /home/osboxes/jupyter/notebook
[D 15:25:58.465 NotebookApp] Paths used for configuration of jupyter_notebook_config: 
    	/etc/jupyter/jupyter_notebook_config.json
[D 15:25:58.465 NotebookApp] Paths used for configuration of jupyter_notebook_config: 
    	/usr/local/etc/jupyter/jupyter_notebook_config.json
[D 15:25:58.466 NotebookApp] Paths used for configuration of jupyter_notebook_config: 
    	/home/osboxes/jupyter/notebook/.tox/etc/jupyter/jupyter_notebook_config.json
[D 15:25:58.466 NotebookApp] Paths used for configuration of jupyter_notebook_config: 
    	/home/osboxes/.jupyter/jupyter_notebook_config.json
[D 15:25:58.476 NotebookApp] initializing json logging
{"written_at": "2020-10-12T20:25:58.650Z", "written_ts": 1602534358650364000, "msg": "Serving notebooks from local directory: /home/osboxes/jupyter/notebook", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2250}
{"written_at": "2020-10-12T20:25:58.650Z", "written_ts": 1602534358650523000, "msg": "Jupyter Notebook 7.0.0.dev0 is running at:", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2250}
...
```

mriedem replied:
Assuming you have `json-logging` installed too...`pip install -e .[json-logging]`. Otherwise you'd hit the ImportError and skip that debug line.

kevin-bates replied:
I see the expected behavior when the package is not installed, etc.  but it does appear there's a difference when the env is used vs. the CLI option.  In the latter case, here are the results, up to the same period you list above.  All of the logging is JSON but I don't see the particular debug message produced in the validation handler - thus I figured it's not useful.  Since it appears in some cases, I suppose it does provide some value - especially since in this hybrid log format case, it would be helpful to know the sudden JSON output is intentional.  My hunch is that you'd see the same mixed results when configured via a config file, whereas CLI options take precedence and, I suppose, are processed prior to the others as things get loaded.

```
$ jupyter notebook --debug --notebook-dir=~/notebooks --NotebookApp.log_json=True
{"written_at": "2020-10-12T21:17:02.042Z", "written_ts": 1602537422042635000, "msg": "Searching ['/Users/kbates/repos/oss/jupyter/notebook', '/Users/kbates/.jupyter', '/opt/anaconda3/envs/notebook-dev/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 179}
{"written_at": "2020-10-12T21:17:02.042Z", "written_ts": 1602537422042879000, "msg": "Looking for jupyter_config in /etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.043Z", "written_ts": 1602537422043005000, "msg": "Looking for jupyter_config in /usr/local/etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.043Z", "written_ts": 1602537422043118000, "msg": "Looking for jupyter_config in /opt/anaconda3/envs/notebook-dev/etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.043Z", "written_ts": 1602537422043214000, "msg": "Looking for jupyter_config in /Users/kbates/.jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.043Z", "written_ts": 1602537422043305000, "msg": "Looking for jupyter_config in /Users/kbates/repos/oss/jupyter/notebook", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.043Z", "written_ts": 1602537422043996000, "msg": "Looking for jupyter_notebook_config in /etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.044Z", "written_ts": 1602537422044112000, "msg": "Looking for jupyter_notebook_config in /usr/local/etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.044Z", "written_ts": 1602537422044218000, "msg": "Looking for jupyter_notebook_config in /opt/anaconda3/envs/notebook-dev/etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.044Z", "written_ts": 1602537422044313000, "msg": "Looking for jupyter_notebook_config in /Users/kbates/.jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.044Z", "written_ts": 1602537422044774000, "msg": "Loaded config file: /Users/kbates/.jupyter/jupyter_notebook_config.py", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 578}
{"written_at": "2020-10-12T21:17:02.044Z", "written_ts": 1602537422044951000, "msg": "Looking for jupyter_notebook_config in /Users/kbates/repos/oss/jupyter/notebook", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.045Z", "written_ts": 1602537422045964000, "msg": "Raising open file limit: soft 2560->4096; hard 9223372036854775807->9223372036854775807", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "notebookapp", "line_no": 1661}
{"written_at": "2020-10-12T21:17:02.049Z", "written_ts": 1602537422049537000, "msg": "Paths used for configuration of jupyter_notebook_config: __/etc/jupyter/jupyter_notebook_config.json", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "config_manager", "line_no": 97}
{"written_at": "2020-10-12T21:17:02.049Z", "written_ts": 1602537422049934000, "msg": "Paths used for configuration of jupyter_notebook_config: __/usr/local/etc/jupyter/jupyter_notebook_config.json", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "config_manager", "line_no": 97}
{"written_at": "2020-10-12T21:17:02.050Z", "written_ts": 1602537422050385000, "msg": "Paths used for configuration of jupyter_notebook_config: __/opt/anaconda3/envs/notebook-dev/etc/jupyter/jupyter_notebook_config.d/elyra.json__/opt/anaconda3/envs/notebook-dev/etc/jupyter/jupyter_notebook_config.d/jupyterlab_git.json__/opt/anaconda3/envs/notebook-dev/etc/jupyter/jupyter_notebook_config.d/nbdime.json__/opt/anaconda3/envs/notebook-dev/etc/jupyter/jupyter_notebook_config.json", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "config_manager", "line_no": 97}
{"written_at": "2020-10-12T21:17:02.051Z", "written_ts": 1602537422051203000, "msg": "Paths used for configuration of jupyter_notebook_config: __/Users/kbates/.jupyter/jupyter_notebook_config.json", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "config_manager", "line_no": 97}
{"written_at": "2020-10-12T21:17:03.096Z", "written_ts": 1602537423096166000, "msg": "Using default logger", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "loader", "line_no": 342}
{"written_at": "2020-10-12T21:17:03.096Z", "written_ts": 1602537423096397000, "msg": "Using default logger", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "loader", "line_no": 342}
{"written_at": "2020-10-12T21:17:03.096Z", "written_ts": 1602537423096496000, "msg": "Using default logger", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "loader", "line_no": 342}
{"written_at": "2020-10-12T21:17:03.096Z", "written_ts": 1602537423096575000, "msg": "Using default logger", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "loader", "line_no": 342}
{"written_at": "2020-10-12T21:17:03.096Z", "written_ts": 1602537423096649000, "msg": "Using default logger", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "loader", "line_no": 342}
{"written_at": "2020-10-12T21:17:03.098Z", "written_ts": 1602537423098998000, "msg": "Serving notebooks from local directory: /Users/kbates/notebooks", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2250}
{"written_at": "2020-10-12T21:17:03.099Z", "written_ts": 1602537423099071000, "msg": "Jupyter Notebook 7.0.0.dev0 is running at:", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2250}
```

kevin-bates reviewed (commented):
hi Matt - this is looking great.  Just had the one comment regarding the debug statement that doesn't seem to get displayed.

kevin-bates commented:
Hi Matt - would you mind merging with the latest master and pushing that commit?  Your changes with legitimate test failures coincided with a general build issue and I'd like to have this PR be 'green' prior to its merge into master.  Sorry for the inconvenience.

kevin-bates reviewed (approved):
Looks good - thank you Matt.

mriedem commented:
> 
> 
> Looks good - thank you Matt.

Thanks for all of your help Kevin.

mriedem commented:
@kevin-bates quick question, do you have any idea of a timeline for the next planned release? Looking at recent releases it looks like you've been doing monthly releases which is cool. My team would like to avoid pulling this change in via a git hash so if there is a release planned on the horizon we'll wait, just wondering.

kevin-bates commented:
Would getting a release out in the next couple of weeks be sufficient - say last week of October/first week of November? Although convenient, I'd rather not set a fixed cadence unless there are other maintainers that can assist.

mriedem commented:
> Would getting a release out in the next couple of weeks be sufficient - say last week of October/first week of November?

Yup, that would be great.
